# How to do a Monte-Carlo simulation of failure times with lsdfailtools 

This notebook shows how to set up the range of parameters and run a given number of simulation on parallel cores

## First let's import the model and the rainfall data

In [2]:
# Importing the model
import lsdfailtools.iverson2000 as iverson
# I'll need that to process the outputs
import pandas as pd
import numpy as np

# rainfall data
df = pd.read_csv("preprocessed_data.csv")

# Setting the depth resolution vector
depths = np.arange(0.2,3.1,0.1)


## Now let's set up the parameter ranges

The function call is a bit chunky but simply add the different minimum and maximum values for eahc parameter and the code will randomly sample values within it.

Note that `D_0`and `K_sat_min` are sampled in log space and the other are uniform. The depth resolution needs to be the same for each siumulation.

In [ ]:
MCrun = iverson.MonteCarlo_Iverson( alpha_min = 0.1, D_0_min = 1e-6,K_sat_min = 1e-8, d_min = 0.5,Iz_over_K_steady_min = 0.1,
      friction_angle_min = 0.2, cohesion_min = 5000, weight_of_water_min = 9800, 
      weight_of_soil_min = 15000, alpha_max = 0.8, D_0_max = 1e-4,K_sat_max = 1e-6, d_max = 3,Iz_over_K_steady_max = 0.8,
      friction_angle_max = 0.5, cohesion_max = 20000, weight_of_water_max = 9801, 
      weight_of_soil_max = 25000, depths = depths)

## Now let's run the simulations

The simulation are run for a given set of precipitation record. The simulation can take a lot of time, it saves the output into a csv every 20 simulations if you need to look at intermediate resulsts.

In [3]:
MCrun.run_MC_failure_test(df["duration_s"].values, df["intensity_mm_sec"].values,
                          n_process = 2, output_name = "test_MC.csv", n_iterations = 100, replace = True)


m is  [Empty DataFrame
Columns: [alpha, D_0, K_sat, d, Iz_over_K_steady, friction_angle, cohesion, weight_of_water, weight_of_soil, time_of_failure]
Index: [], 'test_MC.csv']
m is        alpha       D_0         K_sat         d  Iz_over_K_steady  \
0  0.338835  0.000013  7.367725e-08  0.661763          0.341909   

   friction_angle     cohesion  weight_of_water  weight_of_soil  \
0        0.388999  5004.113259      9800.010386    15788.498986   

   time_of_failure  
0        7602278.5  
m is        alpha       D_0         K_sat         d  Iz_over_K_steady  \
0  0.388451  0.000002  6.968569e-07  2.977544          0.623231   

   friction_angle     cohesion  weight_of_water  weight_of_soil  \
0        0.226972  18480.16953      9800.532306    16341.865111   

   time_of_failure  
0            -9999  
m is        alpha       D_0         K_sat         d  Iz_over_K_steady  \
0  0.726721  0.000002  7.209338e-07  1.559036          0.168466   

   friction_angle     cohesion  weight_of_water 

0            -9999  
m is        alpha       D_0         K_sat         d  Iz_over_K_steady  \
0  0.710772  0.000003  8.514300e-07  1.448985          0.753368   

   friction_angle     cohesion  weight_of_water  weight_of_soil  \
0        0.290254  6447.891989       9800.47108    20081.389365   

   time_of_failure  
0              0.0  
m is        alpha       D_0         K_sat         d  Iz_over_K_steady  \
0  0.705798  0.000006  1.895972e-08  2.627387          0.503557   

   friction_angle      cohesion  weight_of_water  weight_of_soil  \
0        0.399915  13536.486423      9800.957136    20911.547723   

   time_of_failure  
0              0.0  
m is        alpha      D_0         K_sat         d  Iz_over_K_steady  \
0  0.156479  0.00004  6.908958e-08  0.521421          0.471139   

   friction_angle     cohesion  weight_of_water  weight_of_soil  \
0        0.328926  7619.718782      9800.423006    20783.329847   

   time_of_failure  
0        7467279.5  
m is        alpha       D

0            -9999  
m is        alpha       D_0         K_sat        d  Iz_over_K_steady  \
0  0.534017  0.000013  1.105657e-08  2.05374          0.156645   

   friction_angle      cohesion  weight_of_water  weight_of_soil  \
0         0.30631  10797.037655      9800.755106     19854.55783   

   time_of_failure  
0              0.0  
m is        alpha       D_0         K_sat         d  Iz_over_K_steady  \
0  0.283033  0.000003  3.627333e-07  2.356881          0.247325   

   friction_angle    cohesion  weight_of_water  weight_of_soil  \
0        0.217747  7440.56523      9800.738683      20485.5706   

   time_of_failure  
0            -9999  
m is        alpha       D_0         K_sat         d  Iz_over_K_steady  \
0  0.619324  0.000033  4.913325e-07  2.470895          0.610896   

   friction_angle      cohesion  weight_of_water  weight_of_soil  \
0        0.363195  11930.908396      9800.303505    17165.171707   

   time_of_failure  
0        1075498.5  
m is        alpha       D

0        0.493679  5051.9573      9800.962865    22361.004928              0.0  
m is        alpha       D_0         K_sat         d  Iz_over_K_steady  \
0  0.181177  0.000001  3.002036e-07  1.696687          0.636673   

   friction_angle      cohesion  weight_of_water  weight_of_soil  \
0        0.346692  16414.076315      9800.895803    15082.575688   

   time_of_failure  
0            -9999  
m is        alpha       D_0         K_sat         d  Iz_over_K_steady  \
0  0.542526  0.000009  9.942548e-08  1.867785          0.506437   

   friction_angle      cohesion  weight_of_water  weight_of_soil  \
0        0.430065  15257.157727      9800.246289    18799.118754   

   time_of_failure  
0            -9999  
m is        alpha       D_0         K_sat         d  Iz_over_K_steady  \
0  0.772029  0.000001  1.030676e-07  2.154616          0.170933   

   friction_angle     cohesion  weight_of_water  weight_of_soil  \
0        0.200779  14964.13883      9800.179857    23072.065153   

   

## Now Checking the output

Let's load the csv filecontaining the result. The column "time_of_failure" contains the ... time of failure. -9999 Means no failure and 0. means failure from scratch.

In [2]:
res = pd.read_csv("test_MC.csv")
res["time_of_failure"])

SyntaxError: invalid syntax (<ipython-input-2-c537db5ffb25>, line 2)